### ISSUE TITLE
vNEXT Task: Refactor services to have their own registry of data providers

### ISSUE THREAD

**Issue by StephenHodgson** [Microsoft/MixedRealityToolkit-Unity#3170]
# Overview
Currently all of our Devices are registered as extension services in the main configuration profile, which makes them easy to find, but hard to understand which systems they go to and what their purpose is. Ideally we shouldn't be registering devices as extension services, as that extension service profile should only be for custom services.

There are a few things we can do, such as giving each system its own registry, with more specific interface types to adhere to, so we don't get bombarded with a huge list of services to register. 

I'd also like to rename the "Devices" to to "Data Providers". It made sense for the input system at the time, but as we add more systems (like spatial understanding or networking) we want to categorize these modules to better explain what their function is. This will also give devices a clearer distinction than just a "service" but rather something a service relies on and consumes data from.

All folders and namespaces with "Device" will be renamed to "DataProvider".

for example: 

- `Microsoft.MixedReality.Toolkit.Core.DataProvider.<Type>.<API/Platform>`
- `Microsoft.MixedReality.Toolkit.Core.DataProvider.Controller.WindowsMixedReality`
- `Microsoft.MixedReality.Toolkit.Core.DataProvider.SpatialObservers.WindowsMixedReality`
- `Microsoft.MixedReality.Toolkit.Core.DataProvider.Networking.WebRTC`

One of the biggest changes here is that the device managers currently registered for the input system will move from the extension service registry into the controller mapping profile. This will look very similar to the extension service registry, but with a few key differences:

- Users will be able to only register data provider types

For spatial awareness, the same kind of rule would apply where users would only be able to register spatial observer data providers in a similar registry.

As services need be registered by the `MixedRealityToolkit` service locator as it's initialized, we'll need to add an additional loop in each of the core system registration areas where we check the system's profile for registered data providers and register their services.

# TL;DR

- Extension service profile should only be for custom services that users create for their application.
- Devices (DataProviders) will be registered with each system using a more explicit type.

# Requirements

- A data provider is an abstract hardware/software representation of a resource located outside of the application
- Systems should have a registry of data provider implementations and manage those data providers accordingly
- Data providers should never be accessed by other systems, or services directly, and absolutely never by the presentation & application layer
- The systems should have a means of forwarding events, and providing the application/presentation layer with the expected data in a standardized format.
- Data providers will likely have a base interface/implementation that will provide the common data that the system expects from all resources that it will manage
- Data providers should have a single responsibility and do a very specific task, or implement platform specific features

# Acceptance Criteria

- [ ] `Devices` are no longer registered in the extension service profile
- [ ] `Devices` namespaces and folders to be renamed to `DataProviders`
- [ ] `DataProviders` will get a new interface to use: `IMixedRealityDataProvider` which will inherit from `IMixedRealityService`.
    - [ ] The data provider is still registered as as service, but we're being more explicit about the usage of this particular type of service.
    - [ ] Each type will have it's own more explicit interface for the system it will serve.
- [ ] `DataProviders` can be registered in each system profile
- [ ] `DataProvider` services are registered with the service locator after the core system is successfully registered.
- [ ] I can register device managers with the input system via controller mapping profile
- [ ] I can register the spatial observers with the spatial awareness system via spatial awareness profile.

**Comment by davidkline-ms** [comment]
Let's talk through the implications of this change. On the surface it seems like a great idea. Just would like to understand a little more before committing.

Thanks!

**Comment by Yoyozilla** [comment]
related to two other issues david is work on, david to comment.

**Comment by davidkline-ms** [comment]
complete
